src/kernel/arch/x86/paging.zig

```diff
@@ -438,3 +455,51 @@ test "mapDir" {
         checkDirEntry(entry, virt, virt + PAGE_SIZE_4MB, phys, table);
     }
 }
+
+// The labels to jump to after attempting to cause a page fault. This is needed as we don't want to cause an
+// infinite loop by jummping to the same instruction that caused the fault.
+extern var rt_fault_callback: *u32;
+extern var rt_fault_callback2: *u32;
+
+var faulted = false;
+var use_callback2 = false;
+
+fn rt_pageFault(ctx: *arch.InterruptContext) void {
+    faulted = true;
+    // Return to the fault callback
+    ctx.eip = @ptrToInt(&if (use_callback2) rt_fault_callback2 else rt_fault_callback);
```

Comment:
Thinking about this, I don't get how this works. Following the interrupt chain, there needs to be a `iret` to exit the interrupt. So the code following this, (the rest of the code) is in interrupt mode. And the stack will be messed up. Correct me if I am wrong. Do you need to modify the instruction pointer as shouldn't the code return to where the page fault happened. I haven't written a proper page fault handler, but in real code, if a page fault happened, the OS will load the page and then return the previous running code. But idk if there needs to be a `iret`

Reply:
This fault handler is called by one if the `isrX` functions, so it will return to it, which eventually does an `iret`, jumping to the instruction at `ctx.eip`. We don't want to return to where the page fault happened as that would cause an infinite faulting loop, so we have to jump to the next instruction, hence the callbacks.

Reply:
I see, yes, bc ur not accurately modifying the EIP, but the context which will be popped in the iret.

Reply:
Exactly